Possible flake with unit test: TestReceiverReceive

**Describe the bug**
https://prow.knative.dev/view/gcs/knative-prow/pr-logs/pull/knative_eventing/3212/pull-knative-eventing-unit-tests/1266085156708421633

 **Expected behavior**
Test not to flake.

 **To Reproduce**
Just seen this a few times while doing testing.

 **Knative release version**
HEAD

 **Additional context**
Add any other context about the problem here such as proposed priority

Failures from the above tests:

```
{"level":"info","ts":1590694822.310598,"caller":"client/receiver.go:60","msg":"Listening for events on port 49152"}
{"level":"error","ts":1590694822.310925,"logger":"fallback","caller":"client/client.go:204","msg":"Error while opening the inbound connection: listen tcp :49152: bind: address already in use","stacktrace":"github.com/cloudevents/sdk-go/v2/client.(*ceClient).StartReceiver.func2\n\t/Users/vaikas/projects/go/pkg/mod/github.com/cloudevents/sdk-go/v2@v2.0.0/client/client.go:204"}
{"level":"info","ts":1590694822.313632,"caller":"client/receiver.go:63","msg":"Stopping event receiver on port 49152"}
--- FAIL: TestReceiverReceive (0.00s)
    services_test.go:64: 
        	Error Trace:	services_test.go:64
        	            				services_test.go:46
        	Error:      	Received unexpected error:
        	            	Post "http://localhost:49152/": dial tcp [::1]:49152: connect: connection refused
        	Test:       	TestReceiverReceive
    services_test.go:64: 
        	Error Trace:	services_test.go:64
        	            				services_test.go:47
        	Error:      	Received unexpected error:
        	            	Post "http://localhost:49152/": dial tcp [::1]:49152: connect: connection refused
        	Test:       	TestReceiverReceive
    services_test.go:51: 
        	Error Trace:	services_test.go:51
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 0
        	Test:       	TestReceiverReceive
    services_test.go:52: 
        	Error Trace:	services_test.go:52
        	Error:      	Not equal: 
        	            	expected: 2
        	            	actual  : 1
        	Test:       	TestReceiverReceive
{"level":"info","ts":1590694822.314145,"caller":"config/reader.go:40","msg":"Define config file to be taken into account: /Users/vaikas/.config/wathola/config.toml"}
{"level":"info","ts":1590694822.3148758,"caller":"client/receiver.go:60","msg":"Listening for events on port 49163"}
{"level":"error","ts":1590694822.3151438,"logger":"fallback","caller":"client/client.go:204","msg":"Error while opening the inbound connection: listen tcp :49163: bind: address already in use","stacktrace":"github.com/cloudevents/sdk-go/v2/client.(*ceClient).StartReceiver.func2\n\t/Users/vaikas/projects/go/pkg/mod/github.com/cloudevents/sdk-go/v2@v2.0.0/client/client.go:204"}
{"level":"info","ts":1590694822.318367,"caller":"client/receiver.go:63","msg":"Stopping event receiver on port 49163"}
--- FAIL: TestReceiverReceive (0.00s)
    services_test.go:64: 
        	Error Trace:	services_test.go:64
        	            				services_test.go:46
        	Error:      	Received unexpected error:
        	            	Post "http://localhost:49163/": dial tcp [::1]:49163: connect: connection refused
        	Test:       	TestReceiverReceive
    services_test.go:64: 
        	Error Trace:	services_test.go:64
        	            				services_test.go:47
        	Error:      	Received unexpected error:
        	            	Post "http://localhost:49163/": dial tcp [::1]:49163: connect: connection refused
        	Test:       	TestReceiverReceive
    services_test.go:51: 
        	Error Trace:	services_test.go:51
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 0
        	Test:       	TestReceiverReceive
    services_test.go:52: 
        	Error Trace:	services_test.go:52
        	Error:      	Not equal: 
        	            	expected: 2
        	            	actual  : 1
        	Test:       	TestReceiverReceive

```

--- a/test/lib/await.go
+++ b/test/lib/await.go
@@ -62,11 +62,12 @@ func AwaitForAll(log *zap.SugaredLogger) error {
 }
 
 // WaitForReadiness will wait until readiness endpoint reports OK
-func WaitForReadiness(port int) error {
-	return wait.Poll(25*time.Millisecond, 5*time.Minute, func() (done bool, err error) {
+func WaitForReadiness(port int, log *zap.SugaredLogger) error {
+	return wait.PollImmediate(10*time.Millisecond, 5*time.Minute, func() (done bool, err error) {
 		resp, err := http.Get(fmt.Sprintf("http://localhost:%d/healthz", port))
 		if err != nil {
-			return false, err
+			log.Debugf("Error while connecting: %v", err)
+			return false, nil
 		}
 		defer func() {
 			_ = resp.Body.Close()

--- a/test/test_images/wathola-forwarder/main_test.go
+++ b/test/test_images/wathola-forwarder/main_test.go
@@ -32,7 +32,7 @@ func TestForwarderMain(t *testing.T) {
 	config.Instance.Forwarder.Port = port
 	go main()
 	cancel := <-forwarder.Canceling
-	err := lib.WaitForReadiness(port)
+	err := lib.WaitForReadiness(port, config.Log)
 	assert.NoError(t, err)
 	assert.NotNil(t, instance)
 	cancel()

--- a/test/test_images/wathola-receiver/main_test.go
+++ b/test/test_images/wathola-receiver/main_test.go
@@ -33,7 +33,7 @@ func TestReceiverMain(t *testing.T) {
 	config.Instance.Receiver.Port = port
 	go main()
 	cancel := <-receiver.Canceling
-	err := lib.WaitForReadiness(port)
+	err := lib.WaitForReadiness(port, config.Log)
 	assert.NoError(t, err)
 	assert.NotNil(t, instance)
 

--- a/test/upgrade/prober/wathola/receiver/services_test.go
+++ b/test/upgrade/prober/wathola/receiver/services_test.go
@@ -21,6 +21,8 @@ import (
 	cloudevents "github.com/cloudevents/sdk-go/v2"
 	"github.com/phayes/freeport"
 	"github.com/stretchr/testify/assert"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"knative.dev/eventing/test/lib"
 	"knative.dev/eventing/test/upgrade/prober/wathola/config"
 	"knative.dev/eventing/test/upgrade/prober/wathola/event"
 	"knative.dev/eventing/test/upgrade/prober/wathola/sender"
@@ -32,8 +34,9 @@ import (
 
 func TestReceiverReceive(t *testing.T) {
 	// given
-	e := sender.NewCloudEvent(event.Step{Number: 42}, event.StepType)
-	f := sender.NewCloudEvent(event.Finished{Count: 1}, event.FinishedType)
+	e1 := sender.NewCloudEvent(event.Step{Number: 42}, event.StepType)
+	e2 := sender.NewCloudEvent(event.Step{Number: 43}, event.StepType)
+	f := sender.NewCloudEvent(event.Finished{Count: 2}, event.FinishedType)
 
 	instance := New()
 	port := freeport.GetPort()
@@ -41,21 +44,31 @@ func TestReceiverReceive(t *testing.T) {
 	go instance.Receive()
 	cancel := <-Canceling
 	defer cancel()
+	assert.NoError(t, lib.WaitForReadiness(port, config.Log))
 
 	// when
-	sendEvent(t, e, port)
+	sendEvent(t, e1, port)
+	sendEvent(t, e2, port)
 	sendEvent(t, f, port)
 
 	// then
 	rr := instance.(*receiver)
-	assert.Equal(t, 1, rr.step.Count())
+	assert.NoError(t, waitUntilFinished(rr))
+	assert.Equal(t, 2, rr.step.Count())
 	assert.Equal(t, event.Success, rr.finished.State())
 }
 
 func TestMain(m *testing.M) {
-	config.Instance.Receiver.Teardown.Duration = 20 * time.Millisecond
+	config.Instance.Receiver.Teardown.Duration = time.Millisecond
 	exitcode := m.Run()
 	os.Exit(exitcode)
+}
+
+func waitUntilFinished(r *receiver) error {
+	return wait.PollImmediate(5*time.Millisecond, 30*time.Second, func() (done bool, err error) {
+		state := r.finished.State()
+		return state != event.Active, nil
+	})
 }
 
 func sendEvent(t *testing.T, e cloudevents.Event, port int) {

